Migrate blog content to docs

stevhliu opened the pull request:
This PR migrates the [Writing your own inference pipeline](https://huggingface.co/blog/stable_diffusion#writing-your-own-inference-pipeline) to the docs as part of the efforts to consolidate our docs sources. I added some light edits here and there for clarification and included section headers to give it structure.

The biggest thing here for me is how we use the term `pipeline` which I think can be a little ambiguous. To me, the pipeline has a very specific connotation (the `DiffusionPipeline` and all the pipeline classes that inherit from it). Here, we're showing how to use models and a scheduler to replicate what a pipeline does, but we don't actually use the `DiffusionPipeline` class. Maybe we can clarify that when we're talking about a `pipeline` here, we're talking about a complete diffusion system or something like that? Eager to hear what you think! ❤️

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_2477). All of your documentation changes will be reflected on that endpoint.

yiyixuxu reviewed (commented):
This one is a little bit tricky - I think it's really nice to have a post that shows steps in the pipeline and how it uses individual components - 

But it seems in this example we only swapped out the scheduler? (or did I miss anything?) 

Maybe we can brainstorm a more interesting custom pipeline as an example?

yiyixuxu commented:
@stevhliu 

to answer your question about how we use the term `pipeline`, I think what we tried to do in this blogpost is to teach how to create a custom pipeline - a pipeline class inherit from `DiffusionPipeline` 

(e.g. `class CustomPipeline(DiffusionPipeline): ..`)

it shows the steps you should put in `CustomPipeline.__call__` - basically if you put all the code inside in this tutorial inside `CustomPipeline.__call__` you will get a custom pipeline class (and we probably should do that otherwise it's confusing) 

(I'm a little bit confused by this blog post so I might be wrong😂. )

sayakpaul commented:
Thanks for your efforts!

In line with @yiyixuxu's comment (https://github.com/huggingface/diffusers/pull/2477#issuecomment-1442720919), I think the purpose of the blog post was to show the users how they can collate the different components of a `DiffusionPipeline` (i.e., `vae`, `text_encoder`, `unet`, `scheduler`) and build an end-to-end Diffusion system without essentially wrapping it up as a `DiffusionPipeline`. 

I agree with @stevhliu that if we are saying "Writing your own inference pipeline" then we should clarify further and potentially show a `CustomPipeline` (inherited from the `DiffusionPipeline` class). The pipeline could be simple; we don't need to implement something super customized but we can show how easy it is to customize a pipeline when one is sub-classing from `DiffusionPipeline`.

pcuenca reviewed (commented):
I like the doc structure; however, the example is a bit contrived because it was written before we made it much easier to swap schedulers.

Perhaps we could reframe it as "Understand what goes inside a pipeline" rather than "Create your own pipeline". We can keep the same title, but focus on explaining the components rather than "solving" an artificial task.

pcuenca commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -0,0 +1,199 @@
+
+[[open-in-colab]]
+
+🧨 Diffusers allows you to freely exchange models and schedulers in the pipeline to create your own custom pipeline for inference. This guide will show you how to create a custom Stable Diffusion pipeline for inference with the [`LMSDiscreteScheduler`] instead of the default [`PNDMScheduler`].
```
```suggestion
🧨 Diffusers allows you to freely exchange models and schedulers in the pipeline to create your own custom pipeline for inference. This guide will show you the components that make up the Stable Diffusion pipeline so you can learn how to combine them for other tasks.
```

pcuenca commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -0,0 +1,199 @@
+
+## Load pipeline components
+
+The pretrained [`stable-diffusion-v1-4`](https://huggingface.co/CompVis/stable-diffusion-v1-4/tree/main) checkpoint includes all the components required to setup a complete diffusion pipeline. These components are stored in the following folders:
```
Maybe upgrade to `runwayml/stable-diffusion-v1-5`? Or `stabilityai/stable-diffusion-2-1-base`?

pcuenca commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,199 @@
+>>> unet = UNet2DConditionModel.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="unet")
+```
+
+Instead of loading the default [`PNDMScheduler`], load the [`LMSDiscreteScheduler`] and feel free to configure some of the parameters:
````
```suggestion
Instead of loading the default [`PNDMScheduler`], we load the [`LMSDiscreteScheduler`] to show how to plug pieces together:
```

pcuenca commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,199 @@
+>>> vae.to(torch_device)
+>>> text_encoder.to(torch_device)
+>>> unet.to(torch_device)
+```
````
Maybe add:

> The scheduler is not a PyTorch model, so we don't have to move it to the GPU.

pcuenca reviewed (commented):
Small suggestions exploring how it would look like **if** we decide to reframe it as a way to understand the components.

yiyixuxu commented:
I like @pcuenca's suggestion to reframe this tutorial to explain how the pipeline works. If we decide to go in that direction, I think it makes sense to include this part of the blog post as well 
https://huggingface.co/blog/stable_diffusion#how-does-stable-diffusion-work 

and if we decide to go that direction, where would this page located? - it is not "using diffusers"; and it is a little bit advanced so not suitable in "getting started" as well

sayakpaul commented:
>  If we decide to go in that direction, I think it makes sense to include this part of the blog post as well
https://huggingface.co/blog/stable_diffusion#how-does-stable-diffusion-work

We can do it, but I would not prioritize it since blogs are short-lived content. The official documentation is where we are beginning to put more and more effort because they are usually the first lines of truth. 

> and if we decide to go that direction, where would this page located? - it is not "using diffusers"; and it is a little bit advanced so not suitable in "getting started" as well

Good point! Maybe "LOADING & HUB" as we are aiming to show "Understanding what goes inside a pipeline"?

pcuenca commented:
> >  If we decide to go in that direction, I think it makes sense to include this part of the blog post as well
> https://huggingface.co/blog/stable_diffusion#how-does-stable-diffusion-work
> 
> We can do it, but I would not prioritize it since blogs are short-lived content. The official documentation is where we are beginning to put more and more effort because they are usually the first lines of truth. 

I agree! And I think @yiyixuxu (please, correct me if I'm wrong) was actually proposing to include some of that content inside this doc.

> > and if we decide to go that direction, where would this page located? - it is not "using diffusers"; and it is a little bit advanced so not suitable in "getting started" as well
> 
> Good point! Maybe "LOADING & HUB" as we are aiming to show "Understanding what goes inside a pipeline"? 
> 

Yes, sounds a bit advanced for "getting started" maybe. But I think the original title is fine, as I suppose developers wanting to build stuff with diffusers would search for that task. (Maybe we can include a subtitle too about understanding the pipeline components).

stevhliu commented:
Thanks for the clarification @pcuenca! 

I think this would make a lot of sense as a tutorial (see the explanation for creating a new tutorial section [here](https://github.com/huggingface/diffusers/pull/2473#issuecomment-1448610059)) because it'll help users learn more in-depth what goes on in the pipeline. They'll also learn about the library's core components (model/schedulers), and it'll showcase how you can use the library as a toolbox some more.

> If we decide to go in that direction, I think it makes sense to include this part of the blog post as well

We can have a brief version of the blog content here with a link to the full explanatory material, but I wouldn't recommend adding all of it since it's more important for the learner to follow the directions and get a result instead of describing more conceptual things.

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -0,0 +1,199 @@
+
+# Understanding models and schedulers
+
+[[open-in-colab]]
```
Where is the automatically generated colab for this one?

sayakpaul replied:
I guess the action is run after the PR is merged. The notebooks are created here: https://github.com/huggingface/notebooks

Here's an example of an automatically generated notebook: https://github.com/huggingface/notebooks/blob/main/transformers_doc/en/video_classification.ipynb

Example GitHub Action: https://github.com/huggingface/transformers/actions/runs/1564783799

pcuenca replied:
Do we have such tooling in place?

sayakpaul replied:
I think we just need to add this workflow https://github.com/huggingface/transformers/actions/runs/1564783799/workflow. Maybe @stevhliu would know better.

stevhliu replied:
As far as I know, the automatically generated notebook should be created after merging ([this](https://github.com/huggingface/notebooks/blob/main/diffusers_doc/en/pytorch/write_own_pipeline.ipynb) is where it should be once merged)!

pcuenca commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -10,27 +10,34 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
 <Tip>
 
-💡 While you can load the entire checkpoint and all its components by calling the [`StableDiffusionPipeline`] with the [`~DiffusionPipeline.from_pretrained`] method, the goal of this guide is to show how you can pick and choose the individual models and scheduler you want to use to create a custom inference system.
+💡 Read the [How does Stable Diffusion work?](https://huggingface.co/blog/stable_diffusion#how-does-stable-diffusion-work) blog for more information about the UNet and VAE models work.
```
```suggestion
💡 Read the [How does Stable Diffusion work?](https://huggingface.co/blog/stable_diffusion#how-does-stable-diffusion-work) blog for more information about how the UNet and VAE models work.
```

pcuenca commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -10,27 +10,34 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
 [[open-in-colab]]
 
-🧨 Diffusers allows you to freely exchange models and schedulers in the pipeline to create your own custom pipeline for inference. This guide will show you how to create a custom Stable Diffusion pipeline for inference with the [`LMSDiscreteScheduler`] instead of the default [`PNDMScheduler`].
+A diffusion model consists of two main components, the model and the scheduler. The model's primary function is to denoise noisy input samples. The scheduler handles all the details of the denoising process, such as the number of denoising steps and the amount of noise to add. Depending on the context, the scheduler does slightly different things:
```
This can be a bit confusing as the "model" contains other models. We refer to them later as both "components" and "models". Not sure if attempting to clarify helps.

```suggestion
A diffusion model consists of two main components, the model itself and the scheduler. This is a very high-level conceptual view, as the model contains several components that are themselves pre-trained ML models, as we'll see later. From this view, the model's primary function is to denoise noisy input samples. The scheduler handles all the details of the denoising process, such as the number of denoising steps and the amount of noise to add. Depending on the context, the scheduler does slightly different things:
```

pcuenca commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -61,7 +63,7 @@ Instead of loading the default [`PNDMScheduler`], load the [`LMSDiscreteSchedule
 ```
 
-Move the models to a GPU to speed up inference:
+To speed up inference, move the models to a GPU since they have trainable weights, unlike the scheduler:
````
```suggestion
To speed up inference, move the models to a GPU since, unlike the scheduler, they have trainable weights:
```

sayakpaul commented:
> a) build diffusion pipelines using models and schedulers

Let's focus on this as to me, it's one-of-a-kind selling point of Diffusers. Inside that pipeline you get lots of things for minimal effort: 

* Plug and play different compatible schedulers
* Load separate checkpoints for individual components (VAE, Text Encoder, UNet)
* Lots of cool optimizations

All of these collectively take away the cognitive stress from the practitioners so that they focus on the core offering of a method. The examples that build on top of `diffusers` (that @patrickvonplaten included in https://github.com/huggingface/diffusers/pull/2477#issuecomment-1449755631) show this very well. 

I think if we can make this point conceivable from the doc, it will be really valuable for our users.

stevhliu commented:
Thanks for sharing your vision for the doc @patrickvonplaten! I think I've addressed all the feedback (let me know if I've missed anything), except for the outdated code from the blog. If you can give me some pointers on which parts are outdated and what the current way to do things is, that would be super appreciated!

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,266 @@
+>>> ddpm = DDPMPipeline.from_pretrained("google/ddpm-cat-256").to("cuda")
+>>> image = ddpm(num_inference_steps=25).images[0]
+>>> image
+```
````
should we show the image here?

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,266 @@
+    ```py
+    >>> scheduler.set_timesteps(50)
+    ```
+
````
Can we squueze in a new point 3.) here that shows the reader how calling `set_timesteps(50)` has created the `scheduler.timesteps` tensor and what it contains? This would be pretty important to understand IMO.

E..g something like:

```suggestion
3. Having set the number of timesteps, the scheduler has now created a timesteps array we should run over for the denoising process:

```py
>>> sceduler.timesteps
>>> # [999, ...., ....]
```

As one can see this array contains 50 elements spaced evenly from 1000 to 0. The elements correspond to the timesteps at which the model will denoise the image at.

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,266 @@
+    >>> noise = torch.randn((1, 3, sample_size, sample_size)).to("cuda")
+    ```
+
+4. Now write a loop to iterate over the number of timesteps. At each timestep, the model does a [~`UNet2DModel.forward`] pass and returns the noisy residual. The scheduler's [`~DDPMScheduler.step`] method takes the noisy residual, timestep, and input and it predicts the image at the previous timestep. This output becomes the next input to the model in the denoising loop, and it'll repeat until it ...
````
```suggestion
4. Now write a loop to iterate over the timesteps. At each timestep, the model does a [~`UNet2DModel.forward`] pass and returns the noisy residual. The scheduler's [`~DDPMScheduler.step`] method takes the noisy residual, timestep, and input and it predicts the image at the previous timestep. This output becomes the next input to the model in the denoising loop, and it'll repeat until it reaches the end of the `timesteps` array.
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,266 @@
+    >>> image
+    ```
+
+In the next section, you'll put your skills to the test and breakdown the more complex Stable Diffusion pipeline, but it's still built from the same components and follows the same denoising pattern.
````
Nice! I'd maybe only add 3, 4 sentences to the relationship between `set_timesteps` and `timesteps` to make it crystal clear

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -0,0 +1,266 @@
+>>> from diffusers import AutoencoderKL, UNet2DConditionModel, PNDMScheduler
+
+>>> vae = AutoencoderKL.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="vae")
+>>> tokenizer = CLIPTokenizer.from_pretrained("openai/clip-vit-large-patch14")
```
```suggestion
>>> tokenizer = CLIPTokenizer.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="tokenizer")
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -0,0 +1,266 @@
+
+>>> vae = AutoencoderKL.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="vae")
+>>> tokenizer = CLIPTokenizer.from_pretrained("openai/clip-vit-large-patch14")
+>>> text_encoder = CLIPTextModel.from_pretrained("openai/clip-vit-large-patch14")
```
The user shouldn't have to know which clip was used for SD1-4

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,266 @@
+>>> unet = UNet2DConditionModel.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="unet")
+```
+
+Instead of the default [`PNDMScheduler`], exchange it for the [`LMSDiscreteScheduler`] to see how easy it is to plug a different scheduler in:
````
LMS is very outdated, let's maybe use UniPCScheduler instead

patrickvonplaten replied:
```suggestion
Instead of the default [`PNDMScheduler`], exchange it for the [`UniPCMultistepScheduler`] to see how easy it is to plug a different scheduler in:
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,266 @@
+Instead of the default [`PNDMScheduler`], exchange it for the [`LMSDiscreteScheduler`] to see how easy it is to plug a different scheduler in:
+
+```py
+>>> from diffusers import LMSDiscreteScheduler
````
```suggestion
>>> from diffusers import UniPCMultistepScheduler
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,266 @@
+```py
+>>> from diffusers import LMSDiscreteScheduler
+
+>>> scheduler = LMSDiscreteScheduler(
````
```suggestion
>>> scheduler = UniPCMultistepScheduler.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="scheduler")
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -0,0 +1,266 @@
+>>> prompt = ["a photograph of an astronaut riding a horse"]
+>>> height = 512  # default height of Stable Diffusion
+>>> width = 512  # default width of Stable Diffusion
+>>> num_inference_steps = 100  # Number of denoising steps
```
```suggestion
>>> num_inference_steps = 30  # Number of denoising steps
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -0,0 +1,266 @@
+...     prompt, padding="max_length", max_length=tokenizer.model_max_length, truncation=True, return_tensors="pt"
+... )
+
+>>> text_embeddings = text_encoder(text_input.input_ids.to(torch_device))[0]
```
```suggestion
>>> with torch.no_grad():
>>>     text_embeddings = text_encoder(text_input.input_ids.to(torch_device))[0]
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,266 @@
+>>> uncond_embeddings = text_encoder(uncond_input.input_ids.to(torch_device))[0]
+```
+
+For unconditional image generation, you typically do two forward passes, one for each of the embeddings. But in practice, it is better to concatenate the embeddings into a single batch to avoid doing two forward passes:
````
```suggestion
Let's batch the conditional and unconditional embeddings into a batch to avoid doing two forward passes:
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -0,0 +1,266 @@
+>>> latents = torch.randn(
+...     (batch_size, unet.in_channels, height // 8, width // 8),
+...     generator=generator,
+... )
```
```suggestion
... )
>>> # we divide by 8 since the vae model has 3 down-sampling layers
>>> # 2 ** (len(vae.config.block_out_channels) - 1) == 8
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
```diff
@@ -0,0 +1,266 @@
+
+### Denoise the image
+
+Start by scaling the input with the initial noise distribution, *sigma*, the noise scale value:
```
```suggestion
Start by scaling the input with the initial noise distribution, *sigma*, the noise scale value which is required for 
improved schedulers, such as [`UniPCMultistepScheduler`] as noted in ["Elucidating the Design Space of Diffusion-Based Generative Models"](https://arxiv.org/abs/2206.00364).
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,266 @@
+Start by scaling the input with the initial noise distribution, *sigma*, the noise scale value:
+
+```py
+>>> scheduler.set_timesteps(num_inference_steps)
````
```suggestion
```

patrickvonplaten commented on docs/source/en/using-diffusers/write_own_pipeline.mdx:
````diff
@@ -0,0 +1,266 @@
+```py
+>>> scheduler.set_timesteps(num_inference_steps)
+>>> latents = latents * scheduler.init_noise_sigma
+```
````
Next, we can set the number of timesteps:

```py
>>> scheduler.set_timesteps(num_inference_steps)
```

patrickvonplaten reviewed (approved):
Very cool! Left some suggestions on how it can be improved further. Before merging, it'd be really nice if @patil-suraj @williamberman @pcuenca and @yiyixuxu can also have a look!

yiyixuxu commented:
oh wow I think this post looks really epic now!

patrickvonplaten commented:
Thanks a lot!
